pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,423 @@
+    [Input(f"popover-target-{PAGE}-{VIZ_ID}", "n_clicks")],
+    [State(f"popover-{PAGE}-{VIZ_ID}", "is_open")],
+)
+def toggle_popover(n, is_open):
+    if n:
+        return not is_open
+    return is_open
+
+# callback for dynamically changing the graph title
+@callback(
+    Output(f"graph-title-{PAGE}-{VIZ_ID}", "children"),
+    Input(f"window-width-{PAGE}-{VIZ_ID}", "value"),
+)
+
+def graph_title(window_width):
+    title = f"Contributor Importance in {window_width} Month Windows"
+    return title
+
+# callback for contrib-importance-over-time graph
+@callback(
+    Output(f"{PAGE}-{VIZ_ID}", "figure"),
+    Output(f"check-alert-{PAGE}-{VIZ_ID}", "is_open"),
+    [   
+        Input("repo-choices", "data"),
+        Input(f"patterns-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"threshold-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"window-width-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"step-size-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "start_date"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "end_date"),
+    ],
+    background=True,
+)
+
+def create_contrib_importance_over_time_graph(repolist, patterns, threshold, window_width, step_size, start_date, end_date):
+    # main function for all data pre processing
+    cache = cm()
+    df = cache.grabm(func=ctq, repos=repolist)
+    
+    while df is None:
+        time.sleep(1.0)
+        df = cache.grabm(func=ctq, repos=repolist)
+
+    # data ready.
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+    
+    # if the step size is greater than window width raise Alert
+    if step_size > window_width:
+        return dash.no_update, True
+    
+    df_final = process_data(df, 
+                            patterns, 
+                            threshold, 
+                            window_width, 
+                            step_size, 
+                            start_date, 
+                            end_date)
+
+    fig = create_figure(df_final, step_size)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig, False
+
+def calc_cntrb_importance(df, action_type, threshold):
+    # if the df is empty return None
+    if df.empty:
+        return None
+    
+    # if the specified action type is not in the dfs' cols return None
+    if action_type not in df.columns:
+        return None 
+    
+    # sort rows in df based on number of contributions from greatest to least
+    df = df.sort_values(by=action_type, ascending=False)
+
+    # calculate the threshold amount of contributions 
+    thresh_cntrbs = df[action_type].sum() * threshold
+    
+    # initilize running sum of contributors who make up contributor importance
+    cntrb_importance = 0
+
+    # initialize running sum of contributions
+    running_sum = 0
+    
```

Comment:
The more I think about it here is where the nan columns should be dropped. Then the cntrb importance is not skewed